Stratis.Bitcoin.Features.Wallet/WalletSyncManager.cs

```diff
@@ -88,44 +87,47 @@ public virtual void ProcessBlock(Block block)
                     while (this.chain.GetBlock(fork.HashBlock) == null)
                         fork = fork.Previous;
 
-                    Guard.Assert(fork.HashBlock == block.Header.HashPrevBlock);
                     this.walletManager.RemoveBlocks(fork);
                 }
-                else
+
+                ChainedBlock incomingBlock = this.chain.GetBlock(block.GetHash());
+                if (incomingBlock.Height > this.walletTip.Height)
                 {
-                    ChainedBlock incomingBlock = this.chain.GetBlock(block.GetHash());
-                    if (incomingBlock.Height > this.walletTip.Height)
+                    // the wallet is falling behind we need to catch up
+                    var next = this.walletTip;
+                    while(next != incomingBlock)
                     {
-                        // the wallet is falling behind we need to catch up
-                        var next = this.walletTip;
-                        while(next != incomingBlock)
+                        // while the wallet is catching up the entire node will wait
+                        // if a wallet recovers to a date in the past consensus 
+                        // will stop till the wallet is up to date.
+
+                        // TODO: This code should be replaced with a different approach
+                        // similar to BlockStore the wallet should be standalone and not depend on consensus
+                        // the block should be put in a queue and pushed to the wallet in an async way
+                        // if the wallet is behind it will just read blocks from store (or download in case of a pruned node)
+
+                        next = this.chain.GetBlock(next.Height +1);
+                        Block nextblock = null;
+                        while (true) //replace with cancelation token
```

Comment:
'cancellation'

Reply:
ah well in that case I will actually replace with cancellation